llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -206,6 +243,138 @@ AIEWawRegRewriter::getVRegWithCopies(const MachineBasicBlock &MBB) const {
+    std::set<MCPhysReg> &UsedRegs, const PhysRegVec &AdditionalUsedRegs) const {
+
+  for (auto PhysReg : AdditionalUsedRegs) {
+    if (UsedRegs.count(PhysReg))
+      continue;
+
+    bool OccupiedAllAliasRegs = true;
+    for (MCRegAliasIterator AI(MCRegister(PhysReg), TRI, true); AI.isValid();
+         ++AI) {
+      auto AliasReg = *AI;
+      if (PhysReg == AliasReg)
+        continue;
+
+      // If even one sub-register of PhysReg is not used, PhysReg is also not
+      // used
+      if (TRI->getRegSizeInBits(PhysReg, *MRI) >=
+              TRI->getRegSizeInBits(AliasReg, *MRI) &&
+          UsedRegs.count(AliasReg) == 0) {
+        OccupiedAllAliasRegs = false;
+        break;
+      }
+    }
+
+    // If all the smaller register classes of PhysReg are used, PhysReg is also
+    // used, even if it is not directly used
+    if (OccupiedAllAliasRegs) {
+      LLVM_DEBUG(dbgs() << "All subregisters are used, so alias is also used: "
+                        << printReg(PhysReg, TRI, 0, MRI) << "\n");
+      UsedRegs.insert(PhysReg);
+    }
+  }
+}
+
+void AIEWawRegRewriter::accumulateDefRegs(const MachineBasicBlock &MBB,
+                                          std::set<MCPhysReg> &UsedRegs) const {
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand &OP : MI.defs()) {
+      MCPhysReg PhysReg = getAssignedPhysReg(OP.getReg());
+      if (!MCRegister::isPhysicalRegister(PhysReg))
+        continue;
+
+      UsedRegs.insert(PhysReg);
+      LLVM_DEBUG(dbgs() << "Encountered " << printReg(PhysReg, TRI, 0, MRI)
+                        << " \n");
+    }
+  }
+}
+
+/// Remove callee save registers from \param UsedRegs
+void AIEWawRegRewriter::removeCalleeSaveRegs(
+    std::set<MCPhysReg> &UsedRegs) const {
+  LLVM_DEBUG(dbgs() << "Removing Calle Save Regs from UsedRegs:\n");
+  const auto *CSRegs = MRI->getCalleeSavedRegs();
+
+  for (unsigned I = 0; CSRegs[I]; ++I) {
+    for (MCRegAliasIterator AI(MCRegister(CSRegs[I]), TRI, true); AI.isValid();
+         ++AI) {
+      auto AliasReg = *AI;
+      if (UsedRegs.count(AliasReg)) {
+        LLVM_DEBUG(dbgs() << "Already blocked: "
+                          << printReg(AliasReg, TRI, 0, MRI) << " \n");
+        continue;
+      }
+
+      UsedRegs.erase(AliasReg);
+      LLVM_DEBUG(dbgs() << "Erasing " << printReg(AliasReg, TRI, 0, MRI)
+                        << "\n");
+    }
+  }
+
+  LLVM_DEBUG(dbgs() << "\n");
+}
+
+PhysRegVec
+AIEWawRegRewriter::getAdditionalBlockedRegs(const MachineFunction &MF) const {
+  // Only GPR registers have to be handled specially, since callee save
+  // registers should not be used for renaming (could cause spilling) and unused
+  // gpr registers can be used for register spilling to regs instead of memory.
+  // Spilling to regs is more performant, therefore block these unused registers
+  // for register renaming
+  LLVM_DEBUG(dbgs() << "Setting Additional Blocked Registers\n");
+
+  PhysRegVec GPRRegs = getGPRAndAliasRegs(MF);
```

Comment:
Would there be a way to make this whole code more generic? There seems to be a lot of special code for GPRs here. 

Generally, I would expect:
 - We start with a list of interesting registers to rename, I guess that would be vectors and GPRs. I would expect this is the only place where we mention GPRs or VEC registers. The rest of the code should register-class-agnostic
 - We remove registers that are callee-saved
 - We remove registers that are already defined in the loop

And then we iterate over the registers (or rather: RegUnits) that are defined multiple times. For those we try to re-assign them using the list of "free" registers gathered above.